cpp/src/io/parquet/writer_impl.cu

```diff
@@ -448,56 +537,185 @@ class parquet_column_view {
   }
 
  private:
-  // cudf data column
-  column_view _col;
-  column_view _leaf_col;
-
-  // Identifier within set of columns
-  size_t _id        = 0;
-  bool _string_type = false;
-  bool _list_type   = false;
-
-  size_t _type_width     = 0;
-  size_t _row_count      = 0;
-  size_t _data_count     = 0;
-  size_t _null_count     = 0;
-  void const *_data      = nullptr;
-  uint32_t const *_nulls = nullptr;
-  size_type _offset      = 0;
-
-  // parquet-related members
-  std::string _name{};
-  Type _physical_type;
-  ConvertedType _converted_type;
-  statistics_dtype _stats_dtype;
-  int32_t _ts_scale;
-  std::vector<std::string> _path_in_schema;
-
-  // Dictionary-related members
-  bool _dictionary_used = false;
-  rmm::device_vector<uint32_t> _dict_data;
-  rmm::device_vector<uint32_t> _dict_index;
+  // Schema related members
+  schema_tree_node schema_node;
+  std::vector<std::string> path_in_schema;
+  uint8_t _max_def_level = 0;
+  uint8_t _max_rep_level = 0;
+  rmm::device_uvector<uint8_t> _d_nullability;
+
+  column_view cudf_col;
 
   // List-related members
+  bool _is_list;
   rmm::device_uvector<size_type>
     _dremel_offsets;  ///< For each row, the absolute offset into the repetition and definition
                       ///< level vectors. O(num rows)
   rmm::device_uvector<uint8_t> _rep_level;
   rmm::device_uvector<uint8_t> _def_level;
-  std::vector<bool> _nullability;
-  size_type _max_def_level  = -1;
-  size_type _nesting_levels = 0;
+  std::vector<uint8_t> _nullability;
+  size_type _data_count = 0;
 
-  // String-related members
-  rmm::device_buffer _indexes;
-
-  // Decimal-related members
-  int32_t _decimal_scale     = 0;
-  uint8_t _decimal_precision = 0;
+  // Dictionary related members
+  bool _dictionary_used = false;
+  rmm::device_vector<uint32_t> _dict_data;
+  rmm::device_vector<uint32_t> _dict_index;
 };
 
+parquet_column_view::parquet_column_view(schema_tree_node const &schema_node,
+                                         std::vector<schema_tree_node> const &schema_tree,
+                                         rmm::cuda_stream_view stream)
+  : schema_node(schema_node),
+    _d_nullability(0, stream),
+    _dremel_offsets(0, stream),
+    _rep_level(0, stream),
+    _def_level(0, stream)
+{
+  // Construct single inheritance column_view from linked_column_view
+  auto curr_col                           = schema_node.leaf_column.get();
+  column_view single_inheritance_cudf_col = *curr_col;
+  while (curr_col->parent) {
```

Comment:
We marched down to build the schema and we march back up to write it? Why not store the top-most parent for the leaf node so we don't have to march it again?

Reply:
We are not writing the schema here. `parquet_column_view`'s ctor is only going to read the schema and get information that's useful while encoding the data.

The top most parent still contains all its children. We wouldn't be able to get to this leaf. Here, we're using the leaf to march up and convert it to the single child format. So the parent is like this:
```
   S
  / \
 i   f
 ^   ^--- Some other column's leaf
 ⎣__ this column's leaf
```
By having a pointer to this column's leaf we can create 
```
S
|
i
```